# Assignment 1

This assignment will involve the creation of a spellchecking system and an evaluation of its performance. You may use the code snippets provided in Python for completing this or you may use the programming language or environment of your choice

Please start by downloading the corpus `holbrook.txt` from Blackboard

The file consists of lines of text, with one sentence per line. Errors in the line are marked with a `|` as follows

    My siter|sister go|goes to Tonbury .
    
In this case the word 'siter' was corrected to 'sister' and the word 'go' was corrected to 'goes'.

In some places in the corpus two words maybe corrected to a single word or one word to a multiple words. This is denoted in the data using underscores e.g.,

    My Mum goes out some_times|sometimes .
    
For the purpose of this assignment you do not need to separate these words, but instead you may treat them like a single token.

*Note: you may use any functions from NLTK to complete the assignment. It should not be necessary to use other libraries and so please consult with us if your solution involves any other external library. If you use any function from NLTK in Task 6 please include a brief description of this function and how it contributes to your solution.*

## Task 1 (10 Marks)

Write a parser that can read all the lines of the file `holbrook.txt` and print out for each line the original (misspelled) text, the corrected text and the indexes of any changes. The indexes refers to the index of the words in the sentence. In the example given, there is only an error in the 10th word and so the list of indexes is [9]. It is not necessary to analyze where the error occurs inside the word.

Then split your data into a test set of 100 lines and a training set.

In [4]:
#Importing holbrook.txt from local to Colab
import nltk
from google.colab import files
uploaded = files.upload()

Saving holbrook.txt to holbrook.txt


In [5]:
import nltk
nltk.download('punkt')
def parse(sent):
  lorig = []
  lcorrect = []
  indexes = []
  cnt = 0
  for i in sent:
        if '|' in i:
          # Splitting the sentence on '|'
            spt = i.split('|')
            # Previous word to '|' is storing in lorig list.
            lorig.append(spt[0])
            # Next word to '|' is storing in lcorrect list.
            lcorrect.append(spt[1])
            #Noting down the index of error.
            indexes.append(cnt)
        else:
            # If there is no '|' in sentence, sentence is stored in loriginal and lcorrected as it is.
            lorig.append(i)
            lcorrect.append(i)
        cnt = cnt + 1
  #Loading to lorig, lcorrect and index list to dictionary.      
  dictionary = {'corrected': lcorrect, 'original': lorig, 'indexes': indexes}
  return dictionary


lines = open("holbrook.txt").readlines()
print(len(lines))
data = []
# Write your code here

# Reading the txt file in reading mode
corp_read = open("holbrook.txt", "r")
lines = []
for i in corp_read:
  lines.append(i.strip())

print(lines[1])

# Tokenizing the sentences based on Words
word_tok = [nltk.word_tokenize(i) for i in lines]
  
for i in word_tok:
  data.append(parse(i))
#Printing the Data
for i in data:
  print i



#assert(data[2] == {
 #'original': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], 
  #'corrected': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'sister', '.'], 
  #'indexes': [9]
#})

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
1217
NIGEL THRUSH page 48
{'corrected': ['1', '.'], 'original': ['1', '.'], 'indexes': []}
{'corrected': ['NIGEL', 'THRUSH', 'page', '48'], 'original': ['NIGEL', 'THRUSH', 'page', '48'], 'indexes': []}
{'corrected': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'sister', '.'], 'original': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], 'indexes': [9]}
{'corrected': ['My', 'Dad', 'works', 'at', 'Melton', '.'], 'original': ['My', 'Dad', 'works', 'at', 'Melton', '.'], 'indexes': []}
{'corrected': ['My', 'sister', 'goes', 'to', 'Tonbury', '.'], 'original': ['My', 'siter', 'go', 'to', 'Tonbury', '.'], 'indexes': [1, 2]}
{'corrected': ['My', 'Mum', 'goes', 'out', 'sometimes', '.'], 'original': ['My', 'Mum', 'goes', 'out', 'some_times', '.'], 'indexes': [4]}
{'corrected': ['I', 'go', 'to', 'Bridgebrook', 'i', 'go', 'out', 'sometimes'

The counts and assertions given in the following sections are based on splitting the training and test set as follows

In [6]:
test = data[:100]
train = data[100:]

print(test)
print(train)

[{'corrected': ['1', '.'], 'original': ['1', '.'], 'indexes': []}, {'corrected': ['NIGEL', 'THRUSH', 'page', '48'], 'original': ['NIGEL', 'THRUSH', 'page', '48'], 'indexes': []}, {'corrected': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'sister', '.'], 'original': ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], 'indexes': [9]}, {'corrected': ['My', 'Dad', 'works', 'at', 'Melton', '.'], 'original': ['My', 'Dad', 'works', 'at', 'Melton', '.'], 'indexes': []}, {'corrected': ['My', 'sister', 'goes', 'to', 'Tonbury', '.'], 'original': ['My', 'siter', 'go', 'to', 'Tonbury', '.'], 'indexes': [1, 2]}, {'corrected': ['My', 'Mum', 'goes', 'out', 'sometimes', '.'], 'original': ['My', 'Mum', 'goes', 'out', 'some_times', '.'], 'indexes': [4]}, {'corrected': ['I', 'go', 'to', 'Bridgebrook', 'i', 'go', 'out', 'sometimes', 'on', 'Tuesday', 'night', 'i', 'go', 'to', 'Youth', 'club', '.'], 'original': ['I', 'go', 'to', 'Bridgebrook', 'i', 'go', 'out',

In [7]:
from nltk.tokenize import RegexpTokenizer

#Fetching the correct sentences from the training Set
train_correct = [elem['corrected'] for elem in train]

print(train_correct[1:6])

#Fetching the correct sentences from the test Set
test_correct = [elem['corrected'] for elem in test]

print(test_correct[42:45])

#Fetching the original sentences from the test Set
test_original = [elem['original'] for elem in test]

print(test_original[1:6])

# Eliminating special characters from the list.
tokenizer = RegexpTokenizer(r'\w+')
train_correct = [tokenizer.tokenize(" ".join(elem)) for elem in train_correct]
test_original = [tokenizer.tokenize(" ".join(elem)) for elem in test_original]
test_correct = [tokenizer.tokenize(" ".join(elem)) for elem in test_correct]

print(train_correct[1:6])
print(test_correct[42:45])

[['this', 'was', 'a', 'Royal', 'Enfield', '_', '?', '_'], ['there', 'were', 'some', 'Royal', 'Enfield', 'Racing', 'there', 'as', 'well'], ['I', 'saw', 'a', 'Royal', 'Enfield', 'come', 'first', 'it', 'was', 'like', 'mine', '.'], ['there', 'were', 'the', 'new', 'Japanese', 'Honda'], ['they', 'are', 'very', 'fast', 'and', 'getting', 'quite', 'popular', 'in', 'England']]
[['These', 'are', 'the', 'names', 'of', 'them', 'Plain', 'Hunt', ',', 'plain', 'Bob', ',', 'Grandsire', 'Doubles', '.'], ['Grandsire', 'Triples', 'it', 'takes', 'a', 'long', 'time', 'to', 'ring', 'them', '.'], ['Me', 'and', 'my', 'dad', 'bike', 'to', 'Melton', '.']]
[['NIGEL', 'THRUSH', 'page', '48'], ['I', 'have', 'four', 'in', 'my', 'Family', 'Dad', 'Mum', 'and', 'siter', '.'], ['My', 'Dad', 'works', 'at', 'Melton', '.'], ['My', 'siter', 'go', 'to', 'Tonbury', '.'], ['My', 'Mum', 'goes', 'out', 'some_times', '.']]
[['this', 'was', 'a', 'Royal', 'Enfield', '_', '_'], ['there', 'were', 'some', 'Royal', 'Enfield', 'Racing',

## **Task 2** (10 Marks): 
Calculate the frequency (number of occurrences), *ignoring case*, of all words and their unigram probability from the corrected *training* sentences.

*Hint: use `Counter` to implement this so it may be called many times*

In [8]:
import nltk
nltk.download('punkt')
from collections import Counter
from nltk.util import ngrams
from nltk.probability import FreqDist
from __future__ import division


def unigram(word):
    
    word = word.lower()
    corrList = []
    #Create a list of the corrected training data
    for i in train_correct:
       corrList.append(" ".join(i).lower())
    corrList = " ".join(corrList)
    
    token = nltk.word_tokenize(corrList)
    
    nltk.unigrams = ngrams(token,1)
    #Using Counter function to count the number of occurences
    a = Counter(nltk.unigrams)
    print("The frequencies of all words is " , a)
    #Counter returns a dictionary thus fetching the value of the corresponding key.
    for k, v in a.items():
      if ( k == (word,)):
        print(v)
    
unigram('me')


def prob(word):
    
    corrList = []
    #Create a list of the corrected training data
    for i in train_correct:
       corrList.append(" ".join(i).lower())
    corrList = " ".join(corrList)
    token = nltk.word_tokenize(corrList)
    # the frequency of each vocabulary item in the text
    fd = FreqDist(token)
    # total words does this corpus have
    total_freq = fd.N()
    #print(total_freq)
    
    word_freq = fd[word]
        
    #print("The probability of the word is" , (word_freq/total_freq))
    return (word_freq/total_freq)

prob("me")
    
# Test your code with the following
#assert(unigram("me")==87)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
('The frequencies of all words is ', Counter({('the',): 1334, ('and',): 919, ('a',): 615, ('to',): 560, ('i',): 519, ('in',): 317, ('he',): 299, ('they',): 287, ('was',): 277, ('it',): 276, ('of',): 267, ('on',): 214, ('you',): 210, ('went',): 206, ('is',): 178, ('said',): 159, ('my',): 155, ('for',): 151, ('we',): 149, ('she',): 133, ('got',): 130, ('with',): 129, ('when',): 122, ('had',): 120, ('there',): 115, ('have',): 114, ('one',): 113, ('so',): 113, ('that',): 112, ('up',): 110, ('then',): 110, ('out',): 109, ('not',): 106, ('but',): 105, ('at',): 104, ('will',): 100, ('go',): 98, ('her',): 97, ('down',): 95, ('s',): 92, ('do',): 91, ('all',): 90, ('man',): 88, ('his',): 87, ('me',): 87, ('get',): 87, ('were',): 79, ('old',): 74, ('are',): 71, ('see',): 70, ('him',): 68, ('t',): 64, ('like',): 64, ('very',): 64, ('some',): 63, ('them',): 61, ('be',): 61, ('am',): 60, ('

0.004221456645155029

## **Task 3** (15 Marks): 
[Edit distance](https://en.wikipedia.org/wiki/Edit_distance) is a method that calculates how similar two strings are to one another by counting the minimum number of operations required to transform one string into the other. There is a built-in implementation in NLTK that works as follows:


In [9]:
from nltk.metrics.distance import edit_distance

# Edit distance returns the number of changes to transform one word to another
print(edit_distance("hello", "hi"))

4


Write a function that calculates all words with *minimal* edit distance to the misspelled word. You should do this as follows

1. Collect the set of all unique tokens in `train`
2. Find the minimal edit distance, that is the lowest value for the function `edit_distance` between `token` and a word in `train`
3. Output all unique words in `train` that have this same (minimal) `edit_distance` value

*Do not implement edit distance, use the built-in NLTK function `edit_distance`*

In [159]:
from nltk.metrics.distance import edit_distance
nltk.download('averaged_perceptron_tagger')
def get_candidates(tok):
    
    corrList = []
    #Create a list of the corrected training data
    for i in train_correct:
        corrList.append(" ".join(i))

    corrList = " ".join(corrList)
    token = nltk.word_tokenize(corrList)
    # the frequency of each vocabulary item in the text
    fd = FreqDist(token)
    # unique words this corpus have
    unique_freq = fd.keys()
    
    # Calculate distance between the given word and all words in corpus
    s = []
    for i in unique_freq:
        
        t = edit_distance(i,tok)
        
        s.append(t)
    
    # Create a key Value Pair of the distance and the words in the corpus
    dist = dict(zip(unique_freq, s))
    
    dist_sort = sorted(dist.items(), key=lambda x: x[1])
    #print("Sorted Tuples of words :",dist_sort)
    #Find the minimum distance
    k,v = dist_sort[0]
    minimal_dist = v
    keys_min = filter(lambda d: d[1] == minimal_dist, dist_sort)
    final_keys = [x[0] for x in keys_min]
    
    return final_keys
  
#print("The words closest to the passed word using the edit distance function are", get_candidates("minde"))

 
   # return []
res = get_candidates('minde')
print(res)
# Test your code as follows
#assert get_candidates("minde") == ['mind', 'mine']

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
['mind', 'mine']


## Task 4 (15 Marks):

Write a function that takes a (misspelled) sentence and returns the corrected version of that sentence. The system should scan the sentence for words that are not in the dictionary (set of unique words in the training set) and for each word that is not in the dictionary choose a word in the dictionary that has minimal edit distance and has the highest *unigram probability*. 

*Your solution to this should involve `get_candidates`*


In [152]:
corrList = []
#Create a list of the corrected training data
test_corrList = []
for i in test_correct[42:45]:
       test_corrList.append(" ".join(i).lower())
test_corrList = " ".join(test_corrList)

print (test_corrList)
for i in train_correct:
  corrList.append(" ".join(i))

corrList = " ".join(corrList)
token = nltk.word_tokenize(corrList)
# the frequency of each vocabulary item in the text
fd = FreqDist(token)
# unique words this corpus have
unique_freq = fd.keys()

def correct(sentence):
    # Write your code here
    corrected = []
    
    indexes = []
    for i in sentence:
        # If word not in unique words then calculate suggested words with get_candidate() function
        
        if i.lower() not in unique_freq:
          #indexes.append(cnt)
          
          l = len(get_candidates(i))
          
          if l > 1:
                suggestion = get_candidates(i)
                probability = []
                
                # For each suggested word calculate unigram probability
                for sug in suggestion:
                  p = prob(sug)
                  probability.append(p)
                
                #The suggested word with maximum probability
                corrected.append(suggestion[probability.index(max(probability))])
          else:
            # If there is only 1 suggested word then take that word - no need to calculate probabilities
            corrected.append(get_candidates(i)[0])
        else:
          corrected.append(i)
          
    return corrected

correct(['My', 'siter', 'go', 'to', 'Tonbury'])

#assert(correct(["this","whitr","cat"]) = ['this','white','cat'])   

these are the names of them plain hunt plain bob grandsire doubles grandsire triples it takes a long time to ring them me and my dad bike to melton


['My', 'sister', 'go', 'to', 'Tonbury']

## **Task 5** (10 Marks): 
Using the test corpus evaluate the *accuracy* of your method, i.e., how many words from your system's output match the corrected sentence (you should count words that are already spelled correctly and not changed by the system).

In [160]:
def accuracy():
    # Write your code here
    
    class_test_corr = []
    #Create a list of the corrected test data using the classifier created above
    for i in test_original:
      t = correct(i)
      class_test_corr.append(t)
    total = 0
    count = 0
    sent_corr = []
    #Iterating over test Correct Data
    for i in test_correct:
   
        t = len(i)
        #print("lenght" ,t)
        total = 0
        #If there is no words present in a sentence then its an empty list and hence assigning 100% accuracy for that sentence
        if (t == 0):
          sent_corr.append(1)
          count = count +1
        else:
          for j in range(0,len(class_test_corr[count])):
            #If the output of the classifier contains the word present in the Correct Test Set then we add up the total words correctly predicted by classifier
            if (class_test_corr[count][j] in i):
              total = total + 1
              
          #Accuracy is defined as the total no of correct words predicted by the classifier / Total no of words in that sentence.
          acc = total/t
          sent_corr.append(acc)
          count = count +1
          
    #Averaging the accuracy over all the sentences in the Train Corpus.
    print("Average Accuracy of words in each sentence: " , round(sum(sent_corr)/len(sent_corr)*100,4), "%")
    
    

#print(accuracy(test))
accuracy()

('Average Accuracy of words in each sentence: ', 83.2531, '%')


## **Task 6 (35 Marks):**

Consider a modification to your algorithm that would improve the accuracy of the algorithm developed in Task 3 and 4

* You may resources beyond those provided here.
* You must **not use the test data** in this task.
* Provide a short text describing what you intend to do and why. 
* Full marks for this section may be obtained without an implementation, but an implementation is preferred.
* Your implementation should not consist of more than 50 lines of code

Please note this task is marked according to: demonstration of knowledge from the lecutures (10), originality and appropriateness of solution (10), completeness of description (10) and technical correctness (5)


**Below are the changes (Implemented and not Implemented):**

Here we have followed the logic of argmax() based computation where we have computed the likelihood of a correct word from the trained corpus using edit_distance function and then found out the word with max probability which essentially is in the same concept as Naive Bayes Classifier.

Below are the modifications applied by me -

1. I have ignored the numbers . If this is not implemented the number in test data is replaced with most nearest number in training data. Hence POS tagging gives a tag 'CD' . And I have added a check to avoid sending this word to the classifier.

2. Proper Nouns: 
Here I identified the Proper Nouns. My approach for this is pos_tag the sentence and identify the tags. If the tag has 'NNP' then the sentence has proper Nouns and these doesnt need to be grammatically corrected. So I am basically finding edit_distance() of only those words that doesnt have 'NNP' tag.

3. The o'clock word check in the corpus. I have added separate handling for the o'clock check since the spell checker is splitting up the word into 2 words - 'o' and 'clock' and then depending upon the knowledge of the trained classifier updating the word incorrectly. I have added check for this word as well. 

4. For correcting the sentence grammatically we can use the nltk_data grammar rules to parse our sentence. We can define a grammar, or use one that is provided, along with a context-free parser. If the sentence parses, then it has valid grammar; if not, then it doesn't. I have not implemented this in my code below.

5. We can collect more data and validate words with Standard Dictionary:
From observation, if we collect more training data there is a posibility of better results. Also if we use any standard dictionary to validate the words may result in slight increase in accuracy. This approach is not implemented in below as well.

6. Bigrams/Trigrams (Not Impemented):
Bigram/Trigram probability can also be implemented and it may result in slightly better performance. However, most of the sentences are short sentences and hence I believe this probability may not result much increase in performance. However, this needs to be experimented and then observe the results.

7. Smoothing (Not Implemented):
Even though we may not have the chance of zero probability as we are using the edit distance function but can use the below smoothing as well - 
 a. The POS algorithm, if the candidate word has a similar part-of-speech tag as the misspelled word.
 b. The N-gram algorithm, if the misspelled word contains similar character subsets as the candidate word, i.e., "wretch" and "retch".
 
8. After the applied modification my accuracy has increased from 83.25 % to around 88% of my spell checker system.

In [148]:
import nltk
from nltk.tag.sequential import RegexpTagger
nltk.download('maxent_ne_chunker')
nltk.download('words')
nltk.download('averaged_perceptron_tagger')



    
def mod_correct(sentence):
    # Write your code here
    corrected = []    
    indexes = []
    count = 0
    lis = ['NNP' ,'CD', 'DT']
    tag = nltk.pos_tag(sentence)
    #print(tag)
    for i in tag: 
      
        #If the word is 'oclock' we need to add the correct word 'o'clock into the corrected list
        if (i[0] == 'oclock'):
            corrected.append('o\'clock')
        # If word not in unique words then calculate suggested words with get_candidate() function   
        # Also I have added the check for not correcting the Proper Nouns, Determiners, and Numeric Dates as they contribute sigficantly less towards wrong spelling.
        elif (i[0].lower() not in unique_freq and (i[1] not in lis)):
          #indexes.append(cnt)          
          l = len(get_candidates(i[0]))          
          if l > 1:
                suggestion = get_candidates(i[0])
                probability = []
                # For each suggested word calculate unigram probability
                for sug in suggestion:
                  p = prob(sug)
                  #print("Probability" , p , "Word" , sug)
                  probability.append(p)
                
                #The suggested word with maximum probability
                corrected.append(suggestion[probability.index(max(probability))])
          else:
            # If there is only 1 suggested word then take that word - no need to calculate probabilities
            corrected.append(get_candidates(i[0])[0])
        
        else:
          corrected.append(i[0])
          
    return corrected
print(mod_correct(['The', 'body', 'was', 'found', 'at', '7', 'oclock', 'in', 'the', 'morning'])) 


[nltk_data] Downloading package maxent_ne_chunker to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package maxent_ne_chunker is already up-to-date!
[nltk_data] Downloading package words to /root/nltk_data...
[nltk_data]   Package words is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
['The', 'body', 'was', 'found', 'at', '7', "o'clock", 'in', 'the', 'morning']


## **Task 7 (5 Marks):**

Repeat the evaluation (as in Task 5) of your new algorithm and show that it outperforms the algorithm from Task 3 and 4

In [161]:
def mod_accuracy():
    # Write your code here
    testorigList = []
    class_test_modcorr = []
    for i in test_original:
      t = mod_correct(i)
      class_test_modcorr.append(t)
    #print(test_original)
    
    total = 0
    count = 0
    sent_corr = []
    for i in test_correct:
   
        t = len(i)
        total = 0
      #If there is no words present in a sentence then its an empty list and hence assigning 100% accuracy for that sentence
        if (t == 0):
          sent_corr.append(1)
          count = count +1
        else:
      
          for j in range(0,len(class_test_modcorr[count])):
      #If the output of the classifier contains the word present in the Correct Test Set then we add up the total words correctly predicted by classifier or
      #if the classifier contains the word 'o'clock' which also essentially contributes to 1 word.
            if ((class_test_modcorr[count][j] in i) or (class_test_modcorr[count][j] == "o'clock")):
                  #total = total + i.count(class_test_corr[count][j])
              total = total + 1
            
          acc = total/t
          sent_corr.append(acc)
          count = count +1
    print("Average Accuracy of words in each sentence: " , round(sum(sent_corr)/len(sent_corr)*100,4), "%")
#print(accuracy(test))
mod_accuracy()

('Average Accuracy of words in each sentence: ', 87.7165, '%')
